codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java

```diff
@@ -600,7 +604,41 @@ protected void onConnectionError(ChannelHandlerContext ctx, Throwable cause, Htt
      */
     protected void onStreamError(ChannelHandlerContext ctx, @SuppressWarnings("unused") Throwable cause,
                                  StreamException http2Ex) {
-        resetStream(ctx, http2Ex.streamId(), http2Ex.error().code(), ctx.newPromise());
+        int streamId = http2Ex.streamId();
+
+        //if this is caused by reading headers that are too large, send a header with status 431
+        if (http2Ex instanceof Http2Exception.HeaderListSizeException &&
+            ((Http2Exception.HeaderListSizeException) http2Ex).duringDecode() &&
+            connection().isServer()) {
+
+            final Http2Stream stream = connection().stream(streamId);
+
+            // NOTE We have to check to make sure that a stream exists before we send our reply.
+            // We likely always create the stream below as the stream isn't created until the
+            // header block is completely processed.
+            if (stream == null) {
+                try {
+                    encoder.connection().remote().createStream(streamId, true);
+                } catch (Http2Exception e) {
+                    resetUnknownStream(ctx, streamId, http2Ex.error().code(), ctx.newPromise());
+                }
+            }
+
+            handleServerHeaderDecodeSizeError(ctx, streamId);
+        }
+
+        resetStream(ctx, streamId, http2Ex.error().code(), ctx.newPromise());
+    }
+
+    /**
+     * Notifies client that this server has received headers that are larger than what it is
+     * willing to accept. Override to change behavior.
+     *
+     * @param ctx the channel context
+     * @param streamId the id of the stream on which the header was received
+     */
+    protected void handleServerHeaderDecodeSizeError(ChannelHandlerContext ctx, int streamId) {
+        frameWriter().writeHeaders(ctx, streamId, headersTooLarge, 0, true, ctx.newPromise());
```

Comment:
Instead of directly calling the `frameWriter()` we should take care that the state of the stream will permit headers being sent. We should check that the state is appropriate for the initial headers, and if not then send a `RST`. It may be best to put this state checking in an internal method and only call the `send error response when it is ok to send initial headers` if the state allows it.

We should also consider using `encoder().writeHeaders` to make sure all the state is updated appropriately.